source/binding/OperatorExpressions.cpp

```diff
@@ -1498,6 +1501,146 @@ void ReplicationExpression::serializeTo(ASTSerializer& serializer) const {
+        return badExpr(compilation, badResult());
+    }
+
+    if (syntax.sliceSize) {
+        if (isRightToLeft) {
+            const auto& sliceExpr =
+                bind(*syntax.sliceSize, context, BindFlags::AllowDataType | BindFlags::Constant);
+            if (sliceExpr.bad())
+                return badExpr(compilation, badResult());
+            if (sliceExpr.kind == ExpressionKind::DataType) {
+                if (!sliceExpr.type->isFixedSize()) {
+                    auto& diag = context.addDiag(diag::BadStreamSlice, sliceExpr.sourceRange);
+                    if (sliceExpr.type->location)
+                        diag.addNote(diag::NoteDeclarationHere, sliceExpr.type->location);
+                    return badExpr(compilation, badResult());
+                }
+                sliceSize = sliceExpr.type->bitstreamWidth();
+            }
+            else {
+                optional<int32_t> count = context.evalInteger(sliceExpr);
+                if (!count)
+                    return badExpr(compilation, badResult());
+                if (*count <= 0) {
+                    context.requireGtZero(count, sliceExpr.sourceRange);
+                    return badExpr(compilation, badResult());
+                }
+                sliceSize = static_cast<bitwidth_t>(*count);
+            }
+        }
+        else
+            context.addDiag(diag::IgnoredSlice, syntax.sliceSize->sourceRange());
+    }
+    else if (isRightToLeft)
+        sliceSize = 1;
+
+    SmallVectorSized<Expression*, 8> buffer;
+    for (const auto argSyntax : syntax.expressions) {
+        Expression* arg;
+        if (assignmentTarget &&
+            argSyntax->expression->kind == SyntaxKind::StreamingConcatenationExpression)
+            arg = &create(compilation, *argSyntax->expression, context, BindFlags::StreamingAllowed,
+                          assignmentTarget);
+        else
+            arg = &selfDetermined(compilation, *argSyntax->expression, context,
+                                  BindFlags::StreamingAllowed);
+        if (argSyntax->withRange) {
+            // TODO: with array_range_expression
+            context.addDiag(diag::NotYetSupported, argSyntax->withRange->sourceRange());
+            return badExpr(compilation, badResult());
+        }
+
+        if (arg->bad())
+            return badExpr(compilation, badResult());
+        if (argSyntax->expression->kind != SyntaxKind::StreamingConcatenationExpression) {
+            const Type& type = *arg->type;
+            // TODO: first-declared member of untagged union
+            if (!type.isBitstreamType(!assignmentTarget)) {
+                context.addDiag(diag::BadStreamType, arg->sourceRange)
+                    << std::string("expression") << type;
+                return badExpr(compilation, badResult());
+            }
+        }
+
+        buffer.append(arg);
+    }
+
+    // Streaming operator has no meaningful type. Use void as a placeholder.
```

Comment:
This diff adds a ton of special casing all over the place for streaming expressions, mostly because of this decision right here. Is it not possible to have the streaming expression determine its type from the assignment-like context? Then there's no need for a new kind of implicit conversion or special assignment rules. Nested streaming concatenations are cast to a packed array of bit using a bit-stream cast, so it seems like it should be possible to get the type there as well.

Reply:
Initially I planned to create a special type for streaming operators too. The closest type is an unpacked struct type. However, struct types needs to be declared in the same scope to be compatible, needs a name, and no existing code to create a struct type except from a syntax. Because it is still quite different from a struct type, doing so may introduce a lot of special casing in type functions of Type.h/Type.cpp for general purposes. The current approach adds casing only around assignments and conversion, and seemed to be less messy than changing type classes/functions.

They are pros and cons between these two approaches. Please let me know if you think the type approach will be better.

Reply:
For constant expressions, [11.2.1] page 255 says

> Constant expressions can use any of the operators defined in Table 11-1.

and streaming operators are in Table 11-1.

Reply:
What I meant was that you could maybe simplify some other uses if you make top-level concatenations have an actual type, for places where you know the assignment target (which you know because it gets passed into this function). Nested concatenations could still have a placeholder type (the void type works ok for this), but then at least a top-level streaming concat expression has a real type associated with it.